File: android/app/src/main/java/com/slackoff/app/support/NovelCacheDownloadManager.kt
package com.slackoff.app.support

import android.content.Context
import android.content.SharedPreferences
import com.slackoff.app.model.NovelChapterItem
import com.slackoff.app.model.NovelChapterList
import com.slackoff.app.model.novelKey
import com.slackoff.app.network.ApiResult
import com.slackoff.app.network.NovelService
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class NovelCacheRecord(
    val source: String,
    val bookId: String,
    val title: String,
    val author: String? = null,
    val category: String? = null,
    val cachedCount: Int = 0,
    val totalCount: Int = 0,
    val isPaused: Boolean = false,
    val failedCount: Int = 0,
    val updatedAt: Long = 0,
) {
    val key: String get() = novelKey(source, bookId)
    val isComplete: Boolean get() = totalCount > 0 && cachedCount >= totalCount
}

data class NovelCacheDownloadState(
    val cachedCount: Int = 0,
    val totalCount: Int = 0,
    val isCaching: Boolean = false,
    val isPaused: Boolean = false,
    val failedCount: Int = 0,
) {
    val isComplete: Boolean get() = totalCount > 0 && cachedCount >= totalCount
    val message: String?
        get() = when {
            isCaching -> null
            isComplete -> "全书缓存完成"
            isPaused -> "缓存已暂停"
            failedCount > 0 -> "缓存完成，$failedCount 章失败，可继续重试"
            else -> null
        }
}

/** App 级整本缓存任务与缓存列表记录。 */
object NovelCacheDownloadManager {
    private const val PREFS_NAME = "com.slackoff.novel-cache-downloads"
    private const val KEY_RECORDS = "records"

    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false; encodeDefaults = true }
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val jobs = ConcurrentHashMap<String, Job>()
    private var prefs: SharedPreferences? = null
    private var recordsByKey = emptyMap<String, NovelCacheRecord>()

    private val _states = MutableStateFlow<Map<String, NovelCacheDownloadState>>(emptyMap())
    val states: StateFlow<Map<String, NovelCacheDownloadState>> = _states.asStateFlow()
    private val _records = MutableStateFlow<List<NovelCacheRecord>>(emptyList())
    val records: StateFlow<List<NovelCacheRecord>> = _records.asStateFlow()

    fun init(context: Context) {
        if (prefs != null) return
        prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val restored = try {
            json.decodeFromString<List<NovelCacheRecord>>(prefs?.getString(KEY_RECORDS, "[]") ?: "[]")
        } catch (_: Exception) {
            emptyList()
        }
        recordsByKey = restored.associateBy { it.key }
        publishRecords()
        _states.value = restored.associate { record ->
            record.key to NovelCacheDownloadState(
                cachedCount = record.cachedCount,
                totalCount = record.totalCount,
                isPaused = record.isPaused || (!record.isComplete && record.cachedCount > 0),
                failedCount = record.failedCount,
            )
        }
    }

    fun refresh(source: String, bookId: String, totalCount: Int) {
        val key = novelKey(source, bookId)
        if (jobs[key]?.isActive == true) return
        scope.launch {
            val cached = NovelChapterCache.cachedChapterCount(source, bookId).coerceAtMost(totalCount)
            updateState(key) { it.copy(cachedCount = cached, totalCount = totalCount) }
            recordsByKey[key]?.let { updateRecord(it.copy(cachedCount = cached, totalCount = totalCount)) }
        }
    }

    fun start(
        source: String,
        bookId: String,
        title: String,
        author: String?,
        category: String?,
        chapters: List<NovelChapterItem>,
    ) {
        if (chapters.isEmpty()) return
        val key = novelKey(source, bookId)
        if (jobs[key]?.isActive == true) return
        updateRecord(
            (recordsByKey[key] ?: NovelCacheRecord(source, bookId, title, author, category)).copy(
                title = title,
                author = author,
                category = category,
                totalCount = chapters.size,
                isPaused = false,
                failedCount = 0,
                updatedAt = System.currentTimeMillis(),
            )
        )
        scope.launch {
            NovelChapterCache.putCatalog(
                source,
                bookId,
                NovelChapterList(source, bookId, title, chapters.size, 0, 0, chapters.size, chapters),
            )
        }
        launchDownload(key, source, bookId, chapters)
    }

    fun pause(source: String, bookId: String) {
        val key = novelKey(source, bookId)
        jobs[key]?.cancel()
        updateState(key) { it.copy(isCaching = false, isPaused = true) }
        recordsByKey[key]?.let { updateRecord(it.copy(isPaused = true, updatedAt = System.currentTimeMillis())) }
    }

    fun resume(record: NovelCacheRecord) {
        if (jobs[record.key]?.isActive == true) return
        scope.launch {
            val catalog = NovelChapterCache.catalog(record.source, record.bookId) ?: return@launch
            updateRecord(record.copy(isPaused = false, failedCount = 0, updatedAt = System.currentTimeMillis()))
            launchDownload(record.key, record.source, record.bookId, catalog.items)
        }
    }

    private fun launchDownload(key: String, source: String, bookId: String, chapters: List<NovelChapterItem>) {
        val job = scope.launch {
            try {
                val missing = chapters.filter { NovelChapterCache.chapter(source, bookId, it.chapterId) == null }
                val initialCached = chapters.size - missing.size
                updateState(key) { NovelCacheDownloadState(initialCached, chapters.size, missing.isNotEmpty()) }
                recordsByKey[key]?.let { updateRecord(it.copy(cachedCount = initialCached, totalCount = chapters.size, isPaused = false)) }
                if (missing.isEmpty()) return@launch

                val failures = AtomicInteger(0)
                val semaphore = Semaphore(4)
                missing.map { chapter ->
                    async {
                        semaphore.withPermit {
                            when (val result = NovelService.chapter(source, bookId, chapter.chapterId)) {
                                is ApiResult.Success -> {
                                    NovelChapterCache.putChapter(source, bookId, result.data)
                                    updateState(key) { current -> current.copy(cachedCount = current.cachedCount + 1) }
                                    val cached = _states.value[key]?.cachedCount ?: initialCached
                                    recordsByKey[key]?.let { record ->
                                        updateRecord(record.copy(cachedCount = cached, updatedAt = System.currentTimeMillis()))
                                    }
                                }
                                is ApiResult.Error -> failures.incrementAndGet()
                            }
                        }
                    }
                }.awaitAll()
                val cached = NovelChapterCache.cachedChapterCount(source, bookId).coerceAtMost(chapters.size)
                updateState(key) { NovelCacheDownloadState(cached, chapters.size, failedCount = failures.get()) }
                recordsByKey[key]?.let {
                    updateRecord(it.copy(cachedCount = cached, totalCount = chapters.size, failedCount = failures.get(), updatedAt = System.currentTimeMillis()))
                }
            } catch (_: CancellationException) {
                throw CancellationException()
            }
        }
        jobs[key] = job
        job.invokeOnCompletion { jobs.remove(key, job) }
    }

    private fun updateState(key: String, transform: (NovelCacheDownloadState) -> NovelCacheDownloadState) {
        _states.update { current -> current + (key to transform(current[key] ?: NovelCacheDownloadState())) }
    }

    @Synchronized
    private fun updateRecord(record: NovelCacheRecord) {
        recordsByKey = recordsByKey + (record.key to record)
        publishRecords()
        prefs?.edit()?.putString(KEY_RECORDS, json.encodeToString(_records.value))?.apply()
    }

    private fun publishRecords() {
        _records.value = recordsByKey.values.sortedByDescending { it.updatedAt }
    }
}

File: SlackOff/Services/NovelCacheDownloadManager.swift
import Foundation

struct NovelCacheRecord: Codable, Identifiable {
    let source: String
    let bookId: String
    var title: String
    var author: String?
    var category: String?
    var cachedCount: Int
    var totalCount: Int
    var isPaused: Bool
    var failedCount: Int
    var updatedAt: Date

    var id: String { NovelBook.key(source: source, bookId: bookId) }
    var isComplete: Bool { totalCount > 0 && cachedCount >= totalCount }
}

struct NovelCacheDownloadState {
    var cachedCount = 0
    var totalCount = 0
    var isCaching = false
    var isPaused = false
    var failedCount = 0
    var isComplete: Bool { totalCount > 0 && cachedCount >= totalCount }
    var message: String? {
        if isCaching { return nil }
        if isComplete { return "全书缓存完成" }
        if isPaused { return "缓存已暂停" }
        if failedCount > 0 { return "缓存完成，\(failedCount) 章失败，可继续重试" }
        return nil
    }
}

/// App 级整本缓存任务及持久化缓存列表。
@Observable
@MainActor
final class NovelCacheDownloadManager {
    static let shared = NovelCacheDownloadManager()
    private static let recordsKey = "com.slackoff.novel-cache-downloads.records"

    private(set) var states: [String: NovelCacheDownloadState] = [:]
    private(set) var records: [NovelCacheRecord] = []
    private var tasks: [String: Task<Void, Never>] = [:]

    private init() {
        if let data = UserDefaults.standard.data(forKey: Self.recordsKey),
           let decoded = try? JSONDecoder().decode([NovelCacheRecord].self, from: data) {
            records = decoded.sorted { $0.updatedAt > $1.updatedAt }
            states = Dictionary(uniqueKeysWithValues: records.map {
                ($0.id, NovelCacheDownloadState(
                    cachedCount: $0.cachedCount,
                    totalCount: $0.totalCount,
                    isPaused: $0.isPaused || (!$0.isComplete && $0.cachedCount > 0),
                    failedCount: $0.failedCount
                ))
            })
        }
    }

    func state(source: String, bookId: String) -> NovelCacheDownloadState {
        states[NovelBook.key(source: source, bookId: bookId)] ?? NovelCacheDownloadState()
    }

    func refresh(source: String, bookId: String, totalCount: Int) {
        let key = NovelBook.key(source: source, bookId: bookId)
        guard tasks[key] == nil else { return }
        let cached = min(NovelChapterCache.shared.cachedChapterCount(source: source, bookId: bookId), totalCount)
        var state = states[key] ?? NovelCacheDownloadState()
        state.cachedCount = cached
        state.totalCount = totalCount
        states[key] = state
        if var record = records.first(where: { $0.id == key }) {
            record.cachedCount = cached
            record.totalCount = totalCount
            save(record)
        }
    }

    func start(
        source: String,
        bookId: String,
        title: String,
        author: String?,
        category: String?,
        chapters: [NovelChapterItem]
    ) {
        guard !chapters.isEmpty else { return }
        let key = NovelBook.key(source: source, bookId: bookId)
        guard tasks[key] == nil else { return }
        let existing = records.first(where: { $0.id == key })
        save(NovelCacheRecord(
            source: source, bookId: bookId, title: title, author: author, category: category,
            cachedCount: existing?.cachedCount ?? 0, totalCount: chapters.count,
            isPaused: false, failedCount: 0, updatedAt: .now
        ))
        let catalog = NovelChapterList(
            source: source, bookId: bookId, title: title, total: chapters.count,
            offset: 0, limit: 0, returned: chapters.count, chapters: chapters
        )
        NovelChapterCache.shared.storeCatalog(catalog, source: source, bookId: bookId)
        launch(key: key, source: source, bookId: bookId, chapters: chapters)
    }

    func pause(source: String, bookId: String) {
        let key = NovelBook.key(source: source, bookId: bookId)
        tasks[key]?.cancel()
        tasks[key] = nil
        var state = states[key] ?? NovelCacheDownloadState()
        state.isCaching = false
        state.isPaused = true
        states[key] = state
        if var record = records.first(where: { $0.id == key }) {
            record.isPaused = true
            record.updatedAt = .now
            save(record)
        }
    }

    func resume(_ record: NovelCacheRecord) {
        guard tasks[record.id] == nil,
              let catalog = NovelChapterCache.shared.catalog(source: record.source, bookId: record.bookId) else { return }
        var resumed = record
        resumed.isPaused = false
        resumed.failedCount = 0
        resumed.updatedAt = .now
        save(resumed)
        launch(key: record.id, source: record.source, bookId: record.bookId, chapters: catalog.items)
    }

    private func launch(key: String, source: String, bookId: String, chapters: [NovelChapterItem]) {
        tasks[key] = Task { [weak self] in
            guard let self else { return }
            let missing = chapters.filter {
                NovelChapterCache.shared.chapter(source: source, bookId: bookId, chapterId: $0.chapterId) == nil
            }
            states[key] = NovelCacheDownloadState(
                cachedCount: chapters.count - missing.count,
                totalCount: chapters.count,
                isCaching: !missing.isEmpty
            )
            guard !missing.isEmpty else { tasks[key] = nil; return }

            let results = await withTaskGroup(of: Bool.self, returning: [Bool].self) { group in
                var iterator = missing.makeIterator()
                for _ in 0..<min(4, missing.count) {
                    if let chapter = iterator.next() {
                        group.addTask { await Self.cacheChapter(chapter, source: source, bookId: bookId) }
                    }
                }
                var values: [Bool] = []
                while let result = await group.next() {
                    if Task.isCancelled { group.cancelAll(); break }
                    values.append(result)
                    if result {
                        var state = states[key] ?? NovelCacheDownloadState(totalCount: chapters.count)
                        state.cachedCount += 1
                        states[key] = state
                        if var record = records.first(where: { $0.id == key }) {
                            record.cachedCount = state.cachedCount
                            record.updatedAt = .now
                            save(record)
                        }
                    }
                    if let chapter = iterator.next() {
                        group.addTask { await Self.cacheChapter(chapter, source: source, bookId: bookId) }
                    }
                }
                return values
            }
            guard !Task.isCancelled else { return }
            let cached = NovelChapterCache.shared.cachedChapterCount(source: source, bookId: bookId)
            states[key] = NovelCacheDownloadState(
                cachedCount: min(cached, chapters.count), totalCount: chapters.count,
                failedCount: results.filter { !$0 }.count
            )
            if var record = records.first(where: { $0.id == key }) {
                record.cachedCount = min(cached, chapters.count)
                record.failedCount = results.filter { !$0 }.count
                record.updatedAt = .now
                save(record)
            }
            tasks[key] = nil
        }
    }

    private func save(_ record: NovelCacheRecord) {
        records.removeAll { $0.id == record.id }
        records.append(record)
        records.sort { $0.updatedAt > $1.updatedAt }
        if let data = try? JSONEncoder().encode(records) {
            UserDefaults.standard.set(data, forKey: Self.recordsKey)
        }
    }

    private nonisolated static func cacheChapter(_ chapter: NovelChapterItem, source: String, bookId: String) async -> Bool {
        if await NovelChapterCache.shared.chapter(source: source, bookId: bookId, chapterId: chapter.chapterId) != nil { return true }
        guard let body = try? await NovelService.chapter(source: source, bookId: bookId, chapterId: chapter.chapterId) else { return false }
        await NovelChapterCache.shared.storeChapter(body, source: source, bookId: bookId)
        return true
    }
}
